doc/specifications/proposed/rest_resource_embedding.md

```diff
@@ -0,0 +1,173 @@
+ <ObjectStates media-type="application/vnd.ez.api.ContentObjectStates+xml" href="/api/ezp/v2/content/objects/1/objectstates"/>
+</Content>
+```
+
+### The `X-eZ-Embed-Value` request header
+
+Which resources must be embedded is specified using this request header.
+It accepts several resources separated by commas:
+
+`X-eZ-Embed-Value: Content.MainLocation,Content.Owner.Groups`
+
+A resource is referenced by its "path" from the root of the response.
+Resources from an embedded resource can also be embedded:
+
+- `Content.MainLocation`
+- `Content.ContentType`
+- `Content.Owner.Groups`
+
+### Permissions
+If the user doesn't have the required permissions to load an embedded
+object, the response will be untouched, and no error will be thrown.
+
+### HTTP Caching
+Responses will vary based on the embedded responses.
+
+Thanks to HTTP cache multi-tagging, customized responses will expire as
+expected: each embedded object will tag the response with the HTTP cache
+tags it requires.
+
+### Implementation
+
+#### Resource links generation in value object visitors
+Value object visitors don't use the router directly anymore to generate
+links to resources. Instead, they build and visit a `ResourceRouteReference`
+object with the name of the route and the route's parameters:
+
+```php
+class LocationValueObjectVisitor extends ValueObjectVisitor
+{
+  public function visit($generator, $visitor, $location)
+  {
+    // ...
+    
+    $generator->startObjectElement$generator->startObjectElement('Content');
+    $visitor->visitValueObject(
+      new ResourceRouteReference(
+        'ezpublish_rest_loadContent',
+        ['contentId' => $location->contentInfo->contentId]
+    );
+    $generator->endObjectElement('Content');
+  }
+}
+```
+
+#### The `ResourceRouteReference` value object visitor
+This object's visitor extends the `RestResourceLink` visitor.
+
+It uses the router to generate a link based on the `RestResourceLink`
+properties, and invokes the parent's `visit()` method.
+
+#### The `RestResourceLink` value object visitor
+It first generate an `href` attribute, respecting the REST output that
+existed before this feature.
+
+It then uses a `PathExpansionChecker` to test if the current generator path,
+returned by the `Generator::getStackPath()` method, is requested for expansion.
+A `RequestHeaderPathExpansionChecker` uses the request to test if expansion
+is needed.
+
+If it is, a `ValueReferenceLoader` loads the referenced
+value object. The returned value object it is visited and added to generated
+output, inside the current object element.
+
+#### The `ExpansionGenerator`
+The `RestResourceLink` visitor passes an `ExpansionGenerator` when visiting
+the loaded value object.
+
+This OutputGenerator decorates the actual (XML or JSON) generator. It will
+skip the first objectElement and its attributes generated for the embedded
+object, in order to avoid duplicate nodes. In the example above, the `LocationList`
```

Comment:
It would because of how the visitors / generators work.

Embedding works inside an objectElement (`<Content` in an XML output) . 

The `RestResourceReference` is created and visited insite that element, for instance `Content` that is visited will first add the `href` attribute, and then visit, if applicable, the referenced object. 

The embedded object's visitor will add its _own_ root tag. For a `Content`, it will add a new `Content` object Element, making you end up with:

```
<Location>
  <!-- that's the first objetElement, where we optionally embed the reference -->
  <Content href="...">
    <!-- this was generated by the Content visitor -->
    <Content href="...">
      <name>...</name>
    </Content>
  </Content>
  <Children href="/api/ezp/v2/content/locations/1/2/children">
    <LocationList href="/api/ezp/v2/content/locations/1/2/children">
      <Location... />
    </LocationList>
  </Children>
</Location>
```